### PR TITLE
Vim and Bash file updates

### PR DESCRIPTION
**Pull request by mkitt**
Some pretty major updates and organization changes described below.
## Simplify bin, Make, git files and directories

These updates effect pretty much everything but the Vim files.
1. Move the theme files (terminal, Xcode) under a single directory
2. Add an install/uninstall script for Xcode to symlink the theme file
3. Set Makefile tasks to mirror the bin scripts
4. Update Makefile ordering
5. Rename git shortcut msg to empty (type `git empty ‘Force rebuild’)`
6. Add a couple of items to the gitignore and update the sort order
7. Tune the bash EDITOR variables (not sure this is working as expected)
8. Simplify the bash PATH variable
9. Minor formatting
10. Try to use a smaller logo in the Readme
## Upgrade Vim plugins, preferences and key commands
1. Remove the vim-jade plugin
2. Replace search plugin ferret for ctrlsf.vim
3. Update nmap and vmap shortcuts for the new ctrlsf.vim search plugin
4. ~~Add lightweight wrapper around silver searcher with the `:Ag` command~~
5. ~~Add a function to highlight search results (used by `Ag` command)~~
6. Guard against setting format options for macOS’s version of Vim
7. Remove a few Syntastic settings which are no longer being used
8. Set the Syntastic eslint executable to `eslint_d` if it’s installed
9. Normalize use of uppercase `<CR>`
10. Remove an unnecessary call to un-highlight whitespace after clearing
11. Add the `bgc` abbreviation for CSS files
12. Remove the abbreviation for `lorem` (I never use it)
13. Add a Readme note about patching SuperTab and new line insertions
## Revamp the search tools and update the mappings
1. Add vim-grepper for wrapping and handling `Ag`, `Ack`, `GitGrep` `grep`, etc.
2. Remove manual `Ag` commands
3. Remove opening of the quick fix window for `CtrlSF`
4. Correct `Supertab` settings
5. Override jumplist commands with `BufSurfForward/BufSurfBack`
6. Add `g` commands that mirror the function key commands (testing)
7. Sort and update a few items
### New Search tools

The big changes here are related to the search in file tools. These tools are mainly wrappers around `ag`, `ack` or something similar. We were using [wincent/ferret](https://github.com/wincent/ferret) but moved to [dyng/ctrlsf.vim](https://github.com/dyng/ctrlsf.vim) for the primary search tool. CtrlSF can handle a number of scenarios including the standard search and open results in a quick fix window (`<F4>` or `gs`). Or as a search and replace tool (`<F5>` or `gz`) coupled with [terryma/vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors) plugin. The main difference with CtrlSF compared to ferret is searching for an item with spaces. CtrlSF requires you to put these types of searches within quotes i.e. `"search term"`. CtrlSF does differ to `Ag` in some of it's support for command line arguments. The main arguments we use to filter are `-G` and `-Q` which are still supported, so there's not much of a mental jump to make. I highly recommend checking out [dyng/ctrlsf.vim](https://github.com/dyng/ctrlsf.vim) for more usage information.

As a secondary and more pure option I added the [vim-grepper](https://github.com/mhinz/vim-grepper) plugin. It provides a lightweight wrapper around ag (`Ag`), ack (`Ack`), git grep (`GG`), grep (`Grep`). These support all of the normal command line arguments you would provide one of these tools in a unix environment. The only shortcut here is `gS` for `Ag`. These are less used but provide a nice fallback. On larger projects, CtrlSF outperforms this plugin pretty substantially hence why it's the primary search tool.
### Key commands

I added some new key commands which are mirrors of the commands used by the function keys. Just testing these out, but most are mapped to the `g` command and may provide better mnemonics and may feel more Vim like. On my local dotfiles, I've removed the function keys all together to try it out. Let me know what you think?
### Setting up eslint syntastic checker

Running eslint in Vim and Syntastic is painfully slow (and blocking). This is due to node startup times and the loading of all of the imported modules. The [eslint d executable](https://github.com/mantoni/eslint_d.js) reduces this overhead by running a server in the background. It's a huge difference in speed. The other nice thing it does is it loads everything from the current working directories `./node_modules` with a fallback to it's own version of eslint. This means no more global installs of eslint and the plugins we use. Just use whatever is in the project. Literally the only thing you need in your global npm path is `eslint_d` and let Vim start it up for you. Less things in your global npm is a good thing, I definitely suggest running:

```
$ npm install -g eslint_d
```

Note though that you should run `eslint_d stop` after updating anything in your local environment.
### Supertab notes

Take a look at the additions to the Readme concerning Supertab and some manual settings you'll need to apply. meh.
## TLDR

Lots of changes, probably pretty easy to digest looking at the diffs though. If you have any questions or concerns please let me know.

:metal:

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -44,6 +43,10 @@ Vim is setup with [vim-plug][vim-plug] as it's plugin manager. The default
 plugins are enabled within the [.vimrc][vimrc] file. To load localized plugins
 add them to a `~/.vimrc.bundles` file.
 
+#### Patch SuperTab inserts new line bug
+There is an [issue with supertab][st162] that effects our setup. Applying the
+patch listed in the [issue][st162] gets around it for now. [Related][st158].
+
 ## Tips
```
Comment (mkitt) [inline Readme.md:50]: Bummer.

File: dots/vimrc [hunk 2]
```diff
@@ -161,43 +166,59 @@ let g:coffee_compile_vert=1
...
+" The `g` commands (testing)
+nnoremap <silent>gF :vertical wincmd f<CR>
+nnoremap <silent>gL :CtrlPBuffer<CR>
+nnoremap <silent>gl :CtrlP<CR>
+nnoremap gS :Ag<space>
+nmap gs <Plug>CtrlSFQuickfixPrompt
+vmap gs <Plug>CtrlSFQuickfixVwordExec
+nmap gz <Plug>CtrlSFPrompt
+vmap gz <Plug>CtrlSFVwordExec
+nnoremap <silent>gy :NERDTreeToggle<CR>
```
Comment (mkitt) [inline dots/vimrc:207]: I associate `g` with `go`. So `gf` is the native way to jump to file (`go file`) under the cursor, `gF` is our shortcut that does the same except in a vertical split. For `gl` and `gL`, all I could come up with is `go list` which fires up `CtrlP` or `CtrlPBuffer` respectively. The search in file results are mapped to `gs` or `go search` in both normal and visual mode. And I use `gz` as `go zearch` for the search and replace feature of CtrlSF. Saying the term `zearch` in your head is very `R` heavy so it kinda works? And the the `Y` in `gy` looks like a Tree maybe even a NERDTree.

For more great `g` commands type `help g`.

File: bin/install [hunk 3]
```diff
@@ -41,11 +41,18 @@ dots() {
   done
 }
 
+#/ xcode           Installs custom Xcode files
+xcode() {
+  mkdir -pv ~/Library/Developer/Xcode/UserData/FontAndColorThemes
+  ln -sfv "`pwd`/themes/polarized.dvtcolortheme" $HOME/Library/Developer/Xcode/UserData/FontAndColorThemes/polarized.dvtcolortheme
+}
+
```
Comment (mkitt) [inline bin/install:49]: We always had this file but never set it up. This just symlinks it.

File: dots/gitconfig [hunk 4]
```diff
@@ -18,7 +18,7 @@
   br = branch
   co = checkout
   del = !FILES=$(git ls-files --deleted) && git rm $FILES
-  msg = commit --allow-empty -m
+  empty = commit --allow-empty -m
   glog = log --pretty='format:%d %Cgreen%h%Creset %an - %s' --graph
```
Comment (mkitt) [inline dots/gitconfig:22]: Convenient for the WebApp when we want to force a rebuild

```
git empty 'Force rebuild for updated environment variables'
```

File: dots/vimrc [hunk 5]
```diff
@@ -161,43 +166,59 @@ let g:coffee_compile_vert=1
...
 noremap <C-l> <C-w>l
 
+" NERDTree in a buffer (like netrw)
+nnoremap <silent>- :silent edit %:p:h<CR>
+nnoremap <silent>_ :silent edit .<CR>
+
+" Override jumplist commands
+nnoremap <silent><C-i> :BufSurfBack<CR>
+nnoremap <silent><C-o> :BufSurfForward<CR>
+
```
Comment (mkitt) [inline dots/vimrc:185]: This overrides existing behavior. These jumplist never make sense to me until we started using the [ton/vim-bufsurf](https://github.com/ton/vim-bufsurf) plugin. Just testing these to see if it makes sense to override them.

**Comment by mkitt** [comment]
cc @jayzes @rynbyjn

**Comment by jayzes** [comment]
Seems reasonable to me!

**Comment by rynbyjn** [comment]
LGTM :metal:
